autoload/vsnip/session/snippet/node/variable.vim

```diff
@@ -3,15 +3,59 @@ function! vsnip#session#snippet#node#variable#import() abort
 endfunction
 
 let s:Variable = {}
+" @see https://code.visualstudio.com/docs/editor/userdefinedsnippets#_variables
+" TODO: BLOCK_COMMENT_START, BLOCK_COMMENT_END, LINE_COMMENT
+let s:known_variables = {
+      \   'TM_SELECTED_TEXT': { -> vsnip#selected_text()},
+      \   'TM_CURRENT_LINE': { -> getline('.')},
+      \   'TM_CURRENT_WORD': { -> ''},
+      \   'TM_LINE_INDEX': { -> line('.') - 1},
+      \   'TM_LINE_NUMBER': { -> line('.')},
+      \   'TM_FILENAME': { -> expand('%:p:t')},
+      \   'TM_FILENAME_BASE': { -> substitute(expand('%:p:t'), '^\@<!\..*$', '', '')},
+      \   'TM_DIRECTORY': { -> expand('%:p:h:t')},
+      \   'TM_FILEPATH': { -> expand('%:p')},
+      \   'CLIPBOARD': { -> getreg(v:register)},
+      \   'WORKSPACE_NAME': { -> ''},
+      \   'CURRENT_YEAR': { -> strftime('%Y')},
+      \   'CURRENT_YEAR_SHORT': { -> strftime('%y')},
+      \   'CURRENT_MONTH': { -> strftime('%m')},
+      \   'CURRENT_MONTH_NAME': { -> strftime('%B')},
+      \   'CURRENT_MONTH_NAME_SHORT': { -> strftime('%b')},
+      \   'CURRENT_DATE': { -> strftime('%d')},
+      \   'CURRENT_DAY_NAME': { -> strftime('%A')},
+      \   'CURRENT_DAY_NAME_SHORT': { -> strftime('%a')},
+      \   'CURRENT_HOUR': { -> strftime('%H')},
+      \   'CURRENT_MINUTE': { -> strftime('%M')},
+      \   'CURRENT_SECOND': { -> strftime('%S')},
+      \   'BLOCK_COMMENT_START': { -> '/**'},
+      \   'BLOCK_COMMENT_END': { -> '*/'},
+      \   'LINE_COMMENT': { -> '//'},
+      \ }
 
 "
 " new.
 "
 function! s:Variable.new(ast) abort
-  return extend(deepcopy(s:Variable), {
-        \   'type': 'variable',
-        \   'name': a:ast.name,
-        \   'children': vsnip#session#snippet#node#create_from_ast(get(a:ast, 'children', [])),
+  if has_key(s:known_variables, a:ast.name)
+    return extend(deepcopy(s:Variable), {
+          \   'type': 'variable',
+          \   'name': a:ast.name,
+          \   'children': vsnip#session#snippet#node#create_from_ast(get(a:ast, 'children', [])),
+          \ })
+  endif
+
+  " When a variable is unknown (that is, its name isn't defined) the name of the variable is inserted and it is transformed into a placeholder.
+  " @see https://code.visualstudio.com/docs/editor/userdefinedsnippets#_variables
+  let l:children = has_key(a:ast, 'children') ?
+        \ vsnip#session#snippet#node#create_from_ast(a:ast.children) :
+        \ [vsnip#session#snippet#node#create_text(a:ast.name)]
+  return extend(deepcopy(vsnip#session#snippet#node#placeholder#import()), {
```

Comment:
I think `s:Variable.new` should not return `placeholder` because it's confusing.

Instead, it should be returns `{ 'type': 'variable', 'unknown': v:true or v:false, ... }`.

We should handle unknown variable in `s:Snippet.sync`, I think.

What do you think about it?

Reply:
Sure. I think it's confusing too.

I think `s:Snippet.sync` is too fat... So, would you think that it'll be implemented in `vsnip#session#snippet#node#create_from_ast` as following?

```vimscript
function! vsnip#session#snippet#node#create_from_ast(ast) abort
  ...
  if a:ast.type ==# 'variable'
    let l:var = s:Variable.new(a:ast)
    if l:var.unkown
      return s:Placeholder.new_from_var(a:ast)
    else
      return l:var
     endif
  endif
  ...
endfunction
```

Reply:
Hm... this approach looks a bit redundunt for me.

1. create placeholders that incomplete state (tabstop not correct).
2. assign tabstop to the placeholders at s:Snippet.sync.

I'm worried about the placeholders that have incorrect tabstop.

But I agree to `s:Snippet.sync` is fat...
Should we create `s:Snippet.init`?

Reply:
Ok! But I couldn't understand  how to use `s:Snippet.init`....

So, we should implement `variable placeholder` in this PR. And then, Refactoring `s:Snippet.sync` in other PRs. What do you think?

Reply:
Sounds good for me!

Thanks for your contribution!
I'm surprised to your work.